doc/explanation/package-management.md

```diff
@@ -0,0 +1,226 @@
+a solution that satifies the given constrains and installation into one step.
+Dune on the other hand separates the steps of finding a solution and installing. 
+First a solution is found and then packages are installed.
+
+The idea of finding a solution and recording it for later is popular in other
+programming language package managers like NPM and is usually called locking.
+
+:::{note}
+`opam` also supports creating lock files. However, these are not as central to
+the opam workflow as they are in the case of package management in Dune, which
+always requires a set of locked packages.
+:::
+
+In the most general sense, a package lock is just a set of specific packages and
+their versions to be installed.
+
+A Dune lock file extends this to a directory with files that describe the
+dependencies. It includes the package's name and version. Unlike many
+other package managers, the files include a lot of other information as well,
+such as the location of the source archives to download (since there is no
+central location for all archives), the build instructions (since each package
+can use its own way of building), and additional metadata like the system
+packages it depends upon.
+
+The information is stored in a directory (`dune.lock` by default) as separate
+files, because that makes them easier to manage in source control as it
+leads to fewer potential merge conflicts and simplifies review processes.
+Storing additional files like patches are also more elegant this way.
+
+### Package Repository Management
+
+To find a valid solution that allows a project to be built, it is necessary to
+know what packages exist, what versions of these packages exist, and what other
+packages these depend on, etc.
+
+In opam, this information is tracked in a central repository called
+[`opam-repository`](https://github.com/ocaml/opam-repository), which contains all
+the metadata for published packages.
+
+It is managed using Git, and opam typically uses a snapshot to find the
+dependencies when searching for a solution that satisfies the constraints.
+
+Likewise, Dune uses the same repository; however, instead of file snapshots, 
+it uses the Git repository directly. In fact, Dune maintains a shared
+internal cache containing all Git repositories that projects use. The
+advantage is that updates to the metadata are very fast because only the newest
+revisions have to be retrieved. The downside is that for the creation of the
+cache, the entire repository has to be cloned first.
+
+Given a priorisation of fast updates, whenever Dune needs to determine the
+available packages, it will update the repository first. Thus each locking
+process by default will use the newest set of packages available.
+
+However, it is also possible to specify specific revisions of the repositories,
+to get a reproducible solution. Due to using Git, any previous revision of the
+repository can be used by specifying a commit hash.
```

Comment:
Here it is in the context of `opam-repository` so metadata repository or "package repository".

Reply:
Ok, this was not clear, because shortly above, the text says:

> In fact, Dune maintains a shared internal cache containing all Git repositories that projects use.

where "all the repository that projects use" is still only referring to metadata repositories, but it seems like it's talking about all repositories involved, including source repositories.

Reply:
Fair enough, I've updated the text to pull the info into a `note` and explain what is happening.